### PR TITLE
Save highlighted dom into hashmap

### PR DESCRIPTION
**Pull request by dem4ron**
Prevents rerunning highlighting loops, highlights initially before render.

https://user-images.githubusercontent.com/66035744/208077436-f31772a8-1837-49dc-abbf-ed932884aacf.mov

### CODE REVIEW

File: app/javascript/components/editor/TestsPanel.tsx [hunk 1]
```diff
@@ -37,6 +57,7 @@ export const TestsPanel = ({
             {testTab.content}
           </code>
         </pre>
+        <div className={!reusing ? 'hidden' : ''} ref={memoTestRef}></div>
```
Comment (ErikSchierboom) [inline app/javascript/components/editor/TestsPanel.tsx]: In general, using `!` is slightly less readable than its logical inverse:

```suggestion
        <div className={reusing ? '' : 'hidden'} ref={memoTestRef}></div>
```

File: app/javascript/components/editor/TestsPanel.tsx [hunk 2]
```diff
@@ -9,16 +9,32 @@ type TestsPanelProps = {
...
 
-  useEffect(() => {
-    if (!testRef.current) {
+  useLayoutEffect(() => {
+    if (!testRef.current || !memoTestRef.current) {
       return
     }
 
-    highlightAll(testRef.current)
+    if (!(testTab.filename in treeMap)) {
```
Comment (iHiD) [inline app/javascript/components/editor/TestsPanel.tsx:30]: This is quite a step up in complexity so can we add some comments above this that explain both the why (e.g. a summary and a link to this PR would be fine) what's going on code wise pls :)

Comment (ErikSchierboom) [reply app/javascript/components/editor/TestsPanel.tsx:30]: I second this

**Comment by iHiD** [comment]
Thanks Aron. Finally, this needs a rebase! :)

**Review by ErikSchierboom — APPROVED** [review]

**Pull request merged**